### PR TITLE
Implement authentication

### LINKED ISSUE #26: Get authentication from a secured GDB server (login/logout)
Add GraphDBServerClient that depends on GDB APIs for login/logout.
Login sends request to /rest/login/** and returns the jwt user and Authorization header. So we need to attach this header to each request afterwords.
Logout should disallow this header.

Token validity is configured using graphdb.auth.token.validity property and by default is 30 days. When the token expired 409:Authentication session expired is returned. Introduce an option to configure whether we obtain new token for re-auth when the token expired or return the error to the user.

### PR DESCRIPTION
**Pull request by svilenvelikov**
If the library is going to be used against a secured server, then all API calls must be authenticated by sending an http authorization header with a token which is obtained after a call to `rest/login/user_name` with a password provided as a specific header.
In case server requires requests to be authenticated, then in the `ServerClientConfig` and `RepositoryClientConfig` must be configured the `username` and `password` which to be used for the authentication. In case `username` and `password` are provided, then the client assumes that authentication is mandatory and the login with the provided credentials is performed authomatically before the first API call. After a successful login, user details which are received and the auth token are stored in the `AuthenticationService`. From that moment on, an `authorization` header with the token as value is sent with every API call which ensures proper authentication and handling.
When the token expires, then the first API call will be rejected with an http error with status `401`. The client handles this automatically by logging-in the user with the same credentials, updates the stored token and retries the initial API call. This behavior is the default and can be changed if the `ServerClientConfig` or `RepositoryClientConfig` are configured with `keepAlive=false`.

Also fixed test which give UnhandledPromiseRejectionWarning during execution. I've no idea how we missed that one.

Fixes #26

### CODE REVIEW

**Comment by svilenvelikov** [comment]
I've some checks failed which I'll address:
- One code duplication which could be fixed and one which is invalid.
- A small drop in the test coverage which I expected because I struggled to do an integration test the flow: login -> logout -> API invocation (e.g. getRepositoryIDs). I'll continue to search for a way to test this.

File: src/server/server-client.js [hunk 1]
```diff
@@ -24,21 +36,64 @@ class ServerClient {
...
+   * @return {Promise<User>} a promise which resolves with an authenticated
+   * {@link User} instance.
+   * @throws {Error} if username or password are missing or login was
+   * unsuccessful.
+   */
+  login(username, password, keepAlive) {
+    if (!username || !password) {
+      throw new Error('Username and password are required parameters!');
+    }
+    this.setKeepAlive(keepAlive);
```
Comment (mihailradkov) [inline src/server/server-client.js]: Is keep alive actually used currently or is something for the future ?

File: src/server/server-client.js [hunk 2]
```diff
@@ -120,6 +173,95 @@ class ServerClient {
...
+    const startTime = Date.now();
+    return this.httpClient.request(requestBuilder).then((response) => {
+      const executionResponse = new HttpResponse(response, this.httpClient);
+      executionResponse.setElapsedTime(Date.now() - startTime);
+      return executionResponse;
+    }).catch((error) => {
+      const status = error.response ? error.response.status : null;
+      // Unauthorized
+      if (status && status === 401) {
+        return this.relogin(this.loggedUser)
```
Comment (mihailradkov) [inline src/server/server-client.js]: Can we control this with `keepAlive` ?

File: src/server/server-client.js [hunk 3]
```diff
@@ -120,6 +173,95 @@ class ServerClient {
...
+  /**
+   * Performs a logout of logged in user. This effectively removes the stored in
+   * the client user. Every consecutive call against secured server will result
+   * in <code>Unauthorized</code> error with status code <code>401</code>.
+   *
+   * @return {Promise} returns a promise which resolves with undefined.
+   */
+  logout() {
+    this.loggedUser.clearToken();
+    return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (factor5) [reply src/server/server-client.js]: I think all API methods should be async for consistency

File: src/server/server-client.js [hunk 4]
```diff
@@ -91,14 +146,12 @@ class ServerClient {
...
-    let elapsedTime = Date.now();
-    return this.httpClient.request(requestBuilder).then(() => {
-      elapsedTime = Date.now() - elapsedTime;
-      this.logger.info({repoId: id, elapsedTime}, 'Deleted repository');
+    const requestBuilder =
+      HttpRequestBuilder.httpDelete(`${SERVICE_URL}/${id}`);
+    return this.execute(requestBuilder).then((response) => {
+      this.logger.info(ConsoleLogger.getLogPayload(response, {repoId: id}),
+        'Deleted repository');
+      return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (factor5) [reply src/server/server-client.js]: Maybe it's redudant really.

File: src/server/server-client.js (hunk 2 above)
Comment (factor5) [reply src/server/server-client.js]: Seems like I lost the implementation of this somehow.

File: src/server/server-client.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (svilenvelikov) [reply src/server/server-client.js]: Are we OK with this?

[the remark this answers is not in the document]
Comment (mihailradkov) [reply src/server/server-client.js]: Ok

File: src/server/server-client.js [hunk 5]
```diff
@@ -120,6 +134,106 @@ class ServerClient {
...
+      })
+      .then((response) => {
+        const executionResponse = new HttpResponse(response, this.httpClient);
+        executionResponse.setElapsedTime(Date.now() - startTime);
+        return executionResponse;
+      })
+      .catch((error) => {
+        const status = error.response ? error.response.status : null;
+        // Unauthorized
+        if (status && status === 401 && this.config.getKeepAlive()) {
```
Comment (svilenvelikov) [inline src/server/server-client.js]: I now see that I can call again execute directly and remove the `relogin` method.

File: src/logging/console-logger.js [hunk 6]
```diff
@@ -77,6 +77,28 @@ class ConsoleLogger extends Logger {
...
+   * Any additional key-value entries from <code>params</code> will be assigned
+   * in the created payload object.
+   *
+   * @static
+   * @param {HttpResponse} response the HTTP response.
+   * Used to get the execution time and the base URL
+   * @param {object} [params] additional parameters to be appended
+   * @return {object} the constructed payload object for logging
+   */
+  static getLogPayload(response, params = {}) {
```
Comment (mihailradkov) [inline src/logging/console-logger.js]: If we decide to use ConsoleLogger.getLogPayload() lets create an issue to remove logging-utils

File: src/repository/repository-client-config.js [hunk 7]
```diff
@@ -13,14 +13,22 @@ class RepositoryClientConfig {
...
    * @param {number} [writeTimeout]
+   * @param {string} [username] username which should be authenticated
+   * @param {string} [pass] the password to be used
+   * @param {boolean} [keepAlive] if the logged in user should be
+   * reauthenticated after auth token expire. This config has meaning when the
+   * server is secured and username and passwords are provided.
    */
   constructor(endpoints, headers, defaultRDFMimeType, readTimeout,
-      writeTimeout) {
+      writeTimeout, username, pass, keepAlive) {
```
Comment (mihailradkov) [inline src/repository/repository-client-config.js]: I'm not sure if we should support setting everything possible property via the constructor.

File: src/service/authentication-service.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+/**
+ * Service dealing with user authentication in a secured server.
+ *
+ * @author Mihail Radkov
+ * @author Svilen Velikov
+ */
+class AuthenticationService {
+  /**
+   * Instantiates the service with the provided HTTP request executor function.
```
Comment (mihailradkov) [inline src/service/authentication-service.js]: But it is not a function

File: src/service/authentication-service.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+  /**
+   * Performs a login request against secured server with provided username and
+   * password. Upon successful authentication a {@link User} instance is created
+   * with the user data and the auth token and returned to the client.
+   *
+   * @param {string} username is the username of the logged in user
+   * @param {string} pass is the password of the logged in user
+   *
+   * @return {Promise<User>} a promise resolving to an authenticated
```
Comment (mihailradkov) [inline src/service/authentication-service.js]: This is not true, L33 resolves without a user.

File: test/repository/rdf-repository-client-authentication.spec.js [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+          subj: '<http://eunis.eea.europa.eu/countries/AZ>',
+          pred: '<http://eunis.eea.europa.eu/rdf/schema.rdf#population>'
+        })
+        .addAcceptHeader(payload.getResponseType());
+      expect(requestMock).toHaveBeenNthCalledWith(2, expectedAPIRequest);
+    });
+  });
+
+  test('should try relogin after token gets expired', () => {
+    mockClient();
```
Comment (mihailradkov) [inline test/repository/rdf-repository-client-authentication.spec.js]: `mockTokenExipration` or `mockRelogin` idk ?

File: src/server/server-client.js [hunk 11]
```diff
@@ -120,6 +134,75 @@ class ServerClient {
...
+   * result will force re-login for the same user to be executed by default,
+   * unless the client is configured with
+   * <code>ServerClientConfig.keepAlive = false</code>
+   *
+   * @private
+   *
+   * @return {Promise} returns a promise which resolves with undefined.
+   */
+  logout() {
+    return this.authenticationService.logout();
```
Comment (mihailradkov) [inline src/server/server-client.js]: Not sure if we really need it but ok.

File: src/logging/console-logger.js (hunk 6 above)
Comment (svilenvelikov) [reply src/logging/console-logger.js]: No, it's OK with the util. I gave up and removed the method from the console logger. :)
Will push in a while.

File: src/service/authentication-service.js (hunk 8 above)
Comment (svilenvelikov) [reply src/service/authentication-service.js]: Well, technically speaking, it is :) but yeah, thx I'll update the doc.

File: src/server/server-client.js (hunk 11 above)
Comment (svilenvelikov) [reply src/server/server-client.js]: I wasn't sure as well, but left it for now.

**Comment by estathop** [comment]
a dummy code example on how to authenticate between graphdb.js node.js and GraphDB 9.0 would be much appreciated

**Comment by svilenvelikov** [comment]
> a dummy code example on how to authenticate between graphdb.js node.js and GraphDB 9.0 would be much appreciated

Thank you for the suggestion. Code examples will be added in the guide when this feature is ready.

**Comment by estathop** [comment]
is there a way to bypass this absence of the feature and somehow do it in another way ? let's say with pure js or node.js ? is it worth investigating ? I m really new to js & node.js

**Comment by svilenvelikov** [comment]
> is there a way to bypass this absence of the feature and somehow do it in another way ? let's say with pure js or node.js ? is it worth investigating ? I m really new to js & node.js

A POST request can be executed with plain js or any http client (e.g. axios) as mentioned in [this stackoverflow answer](https://stackoverflow.com/a/58731810/615036) and the returned token could be set as a header in  the ServerClientConfig or RepositoryClientConfig to be passed with every request.

File: src/repository/repository-client-config.js (hunk 7 above)
Comment (svilenvelikov) [reply src/repository/repository-client-config.js:23]: I agree, but we also support some fluent interface which allows proper initialization too. We can introduce a builder here as well.

**Comment by teodossidossev** [comment]
rebased to master

**Review by mihailradkov — APPROVED** [review]
You can submit new task(s) to address the constructor arguments. For now, it's okay to be merged.

File: README.md [hunk 12]
```diff
@@ -436,10 +436,31 @@ Repository operations like create, edit, delete, shutdown are not supported by t
 
 ### Authorization
 
-*This is a Work In Progress feature (See [this](https://github.com/Ontotext-AD/graphdb.js/pull/28) pull request)*
-
 If the library is going to be used agains a secured server, then all API calls must be authenticated by sending an http authorization header with a token which is obtained after a call to `rest/login/user_name` with a password provided as a specific header.
+
 In case the server requires that requests should be authenticated, then in the `ServerClientConfig` and `RepositoryClientConfig` must be configured the `username` and `password` which to be used for the authentication. If those are provided, then the client assumes that authentication is mandatory and the login with the provided credentials is performed authomatically before the first API call. After a successful login, user details which are received and the auth token are stored in the `AuthenticationService`. From that moment on, with every API call is sent also an `authorization` header with the token as value.
+```javascript
```
Comment (mihailradkov) [inline README.md:442]: Can add headings (or something) to clarify that the first code example is for creating a server configuration while the second is for repository configuration.

**Review by mihailradkov — APPROVED** [review]

**Pull request merged**